##  Merging Datasets
<I><p style="color: blue; font-size: 16px;">
Lets Start with merging important datasets to put foundation of our graph based research. </p></I>
#### Author @ Hari Pad Bharti 


<p>In this work, I tried to merge 5k Movie Datasets with Grouplens movie datasets and get valuable results. </p>

In [162]:
import pandas as pd
import csv
movie = pd.read_csv('movie_metadata.csv') 

Movie_metadata looks like :-               <i>Reference: Kaggle,</i> www.numbers.com 



In [151]:
movie.head(3)

color   director_name  num_critic_for_reviews  duration  \
0  Color   James Cameron                   723.0     178.0   
1  Color  Gore Verbinski                   302.0     169.0   
2  Color      Sam Mendes                   602.0     148.0   

   director_facebook_likes  actor_3_facebook_likes      actor_2_name  \
0                      0.0                   855.0  Joel David Moore   
1                    563.0                  1000.0     Orlando Bloom   
2                      0.0                   161.0      Rory Kinnear   

   actor_1_facebook_likes        gross                           genres  \
0                  1000.0  760505847.0  Action|Adventure|Fantasy|Sci-Fi   
1                 40000.0  309404152.0         Action|Adventure|Fantasy   
2                 11000.0  200074175.0        Action|Adventure|Thriller   

          ...          num_user_for_reviews language  country  content_rating  \
0         ...                        3054.0  English      USA           PG-13   
1         ...                        1238.0  English      USA           PG-13   
2         ...                         994.0  English       UK           PG-13   

        budget  title_year actor_2_facebook_likes imdb_score  aspect_ratio  \
0  237000000.0      2009.0                  936.0        7.9          1.78   
1  300000000.0      2007.0                 5000.0        7.1          2.35   
2  245000000.0      2015.0                  393.0        6.8          2.35   

  movie_facebook_likes  
0                33000  
1                    0  
2                85000  

[3 rows x 28 columns]

And we have 20 Million grouplens movie datasets with repository of users, movies, genre, geographic location etc. <I> Regerence: Grouplens </I>

<p>Now, movie list looks like</p>

In [163]:
df1=pd.read_csv('mov1.csv', skiprows=0)
df1.head(5)

movieId                               title  \
0        1                    Toy Story (1995)   
1        2                      Jumanji (1995)   
2        3             Grumpier Old Men (1995)   
3        4            Waiting to Exhale (1995)   
4        5  Father of the Bride Part II (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

Now, trimming title to find matchable column 'title'. As we have seen above that movie_metadata had movies as only name, so we will remove year from the title and rename it to 'movie title' for matching and merging dataframes. 

In [153]:
df1.columns.values[1]  #checking value at column 1

'title'

Now Removing year from movie_title

In [164]:
df1['title'] = df1['title'].astype(str).str[:-7]
df1.head()

movieId                        title  \
0        1                    Toy Story   
1        2                      Jumanji   
2        3             Grumpier Old Men   
3        4            Waiting to Exhale   
4        5  Father of the Bride Part II   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

Now Renaming title to movie_title

In [155]:
df1.columns.values[1] = 'movie_title'   # column name changed from title to movie_title
df1.head()

movieId                  movie_title  \
0        1                    Toy Story   
1        2                      Jumanji   
2        3             Grumpier Old Men   
3        4            Waiting to Exhale   
4        5  Father of the Bride Part II   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                   Adventure|Children|Fantasy  
2                               Comedy|Romance  
3                         Comedy|Drama|Romance  
4                                       Comedy

Now, making collection of usable columns from movie metadata for further research:-

In [165]:
columns = ['director_name', 'actor_1_name', 'actor_2_name', 'actor_3_name', 'movie_title', 'plot_keywords']


In [166]:
df = movie[columns]
df.head()

director_name     actor_1_name      actor_2_name          actor_3_name  \
0      James Cameron      CCH Pounder  Joel David Moore             Wes Studi   
1     Gore Verbinski      Johnny Depp     Orlando Bloom        Jack Davenport   
2         Sam Mendes  Christoph Waltz      Rory Kinnear      Stephanie Sigman   
3  Christopher Nolan        Tom Hardy    Christian Bale  Joseph Gordon-Levitt   
4        Doug Walker      Doug Walker        Rob Walker                   NaN   

                                         movie_title  \
0                                            Avatar    
1          Pirates of the Caribbean: At World's End    
2                                           Spectre    
3                             The Dark Knight Rises    
4  Star Wars: Episode VII - The Force Awakens    ...   

                                       plot_keywords  
0             avatar|future|marine|native|paraplegic  
1  goddess|marriage ceremony|marriage proposal|pi...  
2                bomb|espionage|sequel|spy|terrorist  
3  deception|imprisonment|lawlessness|police offi...  
4                                                NaN

Now, we slice special characters from movie_title that came from scrapy. 

In [167]:
df['movie_title'] = df['movie_title'].astype(str).str[:-2]
#df['movie_title']

/Users/haripadbharti/anaconda/envs/gl-env/lib/python2.7/site-packages/ipykernel/__main__.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  if __name__ == '__main__':


Now, we merge movie_metadata with grouplens movie data

In [168]:
#df1['title'] = df1['title'].astype(str).str[:-7]
df1['movie_title'] = df1['title']

In [169]:
df3 = pd.merge(df1, df, on='movie_title')

In [171]:
df3.head()

movieId             title                                       genres  \
0        1         Toy Story  Adventure|Animation|Children|Comedy|Fantasy   
1       10         GoldenEye                    Action|Adventure|Thriller   
2       14             Nixon                                        Drama   
3       15  Cutthroat Island                     Action|Adventure|Romance   
4       16            Casino                                  Crime|Drama   

        movie_title    director_name           actor_1_name  \
0         Toy Story    John Lasseter              Tom Hanks   
1         GoldenEye  Martin Campbell      Izabella Scorupco   
2             Nixon     Oliver Stone        Anthony Hopkins   
3  Cutthroat Island     Renny Harlin  Christopher Masterson   
4            Casino  Martin Scorsese         Robert De Niro   

        actor_2_name    actor_3_name  \
0  John Ratzenberger      Jim Varney   
1      Joe Don Baker    Tchéky Karyo   
2        Bob Hoskins      Joan Allen   
3     Frank Langella  Matthew Modine   
4        Don Rickles    Kevin Pollak   

                                       plot_keywords  
0             claw crane|cowboy|jealousy|rivalry|toy  
1  agent|cossack|official james bond series|russi...  
2       courtship|election|paranoia|president|quaker  
3     latin|pirate|pirate ship|treasure|treasure map  
4  car bomb|card cheat|cash|casino|gambling syndi...

In [174]:
df3.count()

movieId          3525
title            3525
genres           3525
movie_title      3525
director_name    3525
actor_1_name     3524
actor_2_name     3523
actor_3_name     3521
plot_keywords    3503
dtype: int64

Now, deleting one extra title in our merged Data Frame df3

In [175]:
del df3['title']

In [176]:
df3.head()

movieId                                       genres       movie_title  \
0        1  Adventure|Animation|Children|Comedy|Fantasy         Toy Story   
1       10                    Action|Adventure|Thriller         GoldenEye   
2       14                                        Drama             Nixon   
3       15                     Action|Adventure|Romance  Cutthroat Island   
4       16                                  Crime|Drama            Casino   

     director_name           actor_1_name       actor_2_name    actor_3_name  \
0    John Lasseter              Tom Hanks  John Ratzenberger      Jim Varney   
1  Martin Campbell      Izabella Scorupco      Joe Don Baker    Tchéky Karyo   
2     Oliver Stone        Anthony Hopkins        Bob Hoskins      Joan Allen   
3     Renny Harlin  Christopher Masterson     Frank Langella  Matthew Modine   
4  Martin Scorsese         Robert De Niro        Don Rickles    Kevin Pollak   

                                       plot_keywords  
0             claw crane|cowboy|jealousy|rivalry|toy  
1  agent|cossack|official james bond series|russi...  
2       courtship|election|paranoia|president|quaker  
3     latin|pirate|pirate ship|treasure|treasure map  
4  car bomb|card cheat|cash|casino|gambling syndi...

In [179]:
#df3.dropna()
df3.count()

movieId          3525
genres           3525
movie_title      3525
director_name    3525
actor_1_name     3524
actor_2_name     3523
actor_3_name     3521
plot_keywords    3503
dtype: int64

<p style="color: black; font-size: 16px;">
At this point of time, we have managed to merge Datasets collected from different sources together to make Actor, Director and Movie list with ID that can potentially match Grouplens data having User Informations, Movie Ratings and other features needed in creating a graph model.
</p>
<p style="color: black; font-size: 18px;">
<b>Note:</b> We are using Grouplens 20Million Dataset and have large dataset out of which, we now have 3525 Movies with Actor, Director and Keyword list. We can also expand our research on online popularity of the movies by analysing our collected data from www.numbers.com using scrapy as available on Kaggle public datasets for R&D purpose.</p>



<p style="color: red; font-size: 18px;">
------------------------------------- Thank You -------------------------------------
</p>